lightning-invoice/src/utils.rs

```diff
@@ -267,29 +268,47 @@ where
 	// the hints across our real nodes we add one hint from each in turn until no node has any hints
 	// left (if one node has more hints than any other, these will accumulate at the end of the
 	// vector).
-	let mut invoice_hints: Vec<RouteHint> = Vec::new();
-	let mut hint_idx = 0;
+	const MAX_HINTS: usize = 3;
 
-	loop {
-		let mut remaining_hints = false;
+    zip_nested_vectors(phantom_hints, MAX_HINTS)
 
-		for hints in phantom_hints.iter() {
-			if invoice_hints.len() == 3 {
-				return invoice_hints
-			}
-
-			if hint_idx < hints.len() {
-				invoice_hints.push(hints[hint_idx].clone());
-				remaining_hints = true
-			}
-		}
-
-		if !remaining_hints {
-			return invoice_hints
-		}
+}
 
-		hint_idx +=1;
-	}
+// Zip multiple nested vectors
+// Similar to the zip method on a vector, it combines the vectors by index - zero index comes first
+// then second, etc.
+// The difference is that this function handles many or just one vector, whereas the builtin method
+// works on just two vectors.
+fn zip_nested_vectors<T: Clone>(vecs: Vec<Vec<T>>, result_size: usize) -> impl Iterator<Item = T> {
+    let max_vector_length: usize = vecs.iter().map(|x| x.len()).max().unwrap();
+    let mut hint_index = 0;
+    let mut vector_index = 0;
+    let number_inner_vectors: usize = vecs.len();
+    let mut total_hints_returned = 0;
+
+    core::iter::from_fn(move || loop {
```

Comment:
Why is the outer `loop` needed?

Reply:
Good point.  I can remove it.  At first, I thought I needed it to manage the logic of what value to retrieve.  But you are right, I can simply return with `None` or `hint_value` (replace `return` with `break`)